### ISSUE TITLE
System.IO.Directory.Move claims to throw when the target already exists; however it does not do so reliably.

### ISSUE THREAD

**Issue by joshudson** [dotnet/runtime#31228]
> Starting with .NET Core 3.0, the Move method throws an IOException in all platforms when the destDirName already exists. In .NET Core 2.2 and previous versions, the exception was only thrown on Windows, and other platforms could either fail or overwrite the destDirName. See C++ rename.

Unfortunately, the fix was made by inserting a call to `FileExists()`, which is broken. Even if `FileExists()` worked, this is subject to a TOCTOU bug. As a consequence, rename targets can be randomly overwritten.

This can be actually fixed in Linux by using the `renameat2()` system call; however I could find no solution for OSX.

This problem can also occur on Windows when talking to set-top NAS boxes. Two computers may be required to reproduce.

**Comment by JeremyKuhne** [comment]
Triage:  Any prototypes to improve this are welcome. We also can consider improvements for non-OSX platforms. Performance is a consideration, of course.

**Comment by joshudson** [comment]
I can fix this for Linux immediately. I can't fix the Windows case for set-top NAS boxes. Better to just document the race condition?

**Comment by joshudson** [comment]
Working on this, I found the following in File.Move():

            Interop.Sys.FileStatus sourceStat, destStat;
            if (Interop.Sys.LStat(sourceFullPath, out sourceStat) == 0 && // source file exists
               (Interop.Sys.LStat(destFullPath, out destStat) != 0 || // dest file does not exist
                    sourceStat.Ino == destStat.Ino) && // source and dest are the same file on that device

The only way this `if` statement can ever be true is if you have memory corruption, or the jitter gets smart enough to elide the zero-initialization of `destStat` since it's initialized by `out` (at which point if we still supported CAS I'd be filing a security bug, but we don't).

I'm taking this block out anyway because I need to push the test down to native code.

**Comment by joshudson** [comment]
Well runtime doesn't build out of the box on buster yet. I've got a not-ready-yet checkpoint while I work on this if anybody wants a preview of the fix.
[checkpoint.zip](https://github.com/dotnet/runtime/files/4302663/checkpoint.zip)

**Comment by danmosemsft** [comment]
Is there an issue for not building on Buster @joshudson

**Comment by joshudson** [comment]
@danmosemsft : Nope. Looks intentional. Somebody upgrade the cmake version demand, and buster doesn't have a high enough version yet.

**Comment by joshudson** [comment]
So I finally figured out what that if block above is actually doing. I had somehow managed to exclude it from consideration because of how broken the behavior is. "if the destination doesn't exist try the rename"; thus completely defeating the point of the link and remove operation below. 

So this issue exists for File.Move as well as Directory.Move.

**Comment by am11** [comment]
> buster doesn't have a high enough version yet.

We can install cmake from official source on any Linux, e.g.

```sh
curl -O https://github.com/Kitware/CMake/releases/download/v3.16.5/cmake-3.16.5-Linux-x86_64.sh

sh cmake-3.16.5-Linux-x86_64.sh --skip-license
```

**Comment by danmosemsft** [comment]
@jkotas @jkoritzinsky should we be supporting build on Buster?

**Comment by jkotas** [comment]
The prerequisites for building are documented at https://github.com/dotnet/runtime/blob/master/docs/workflow/requirements/linux-requirements.md#toolchain-setup .

**Comment by joshudson** [comment]
@jkotas: I'm not absolutely certain because I'm having a hard time reading the azure pipeline output but I think you need to update your build machines. It would appear that `stdio.h` is out-of-date. I know what's causing the Mac OSX failure and I'm working on it. I had hoped that I'd be able to solve the race condition for `root` on Mac OSX but I'm going to have to sacrifice that so that renaming directories with a trailing `/` works instead as there's a test case for it.

**Comment by jkotas** [comment]
> I think you need to update your build machines.

Our official builds are on CentOS 7. We are intentionally using this CentOS 7 for official builds because of it allows us to produce binaries that work everywhere.

**Comment by joshudson** [comment]
@jkotas: Is the minimum Linux kernel version >= 3.15? If so, this can be solved by some more build-time magic.

**Comment by jkotas** [comment]
@janvorli What is the minimum kernel version that we need?

**Comment by joshudson** [comment]
As it stands, the PR accomplishes the following:

1) The corresponding race condition for File.Move() is removed assuming at actually was a file and not a directory and hard links can be used.
2) File.Move() with an actual file is faster due to requiring two fewer managed->native transitions. The system call count required is the same.
3) Calling Directory.Move() is faster due to requiring one fewer managed->native transitions.
4) Calling Directory.Move() to change the case of a directory on a case sensitive file system now works again.
5) Directory.Move() no longer clobbers a file by the name of a target in the face of a transient network glitch (network-filesystem*).
6) The PR adds latent code to fix the race condition for both files and directories for almost all Linux local filesystems. The code is opportunistically activated whenever glibc version is high enough. It would be trivial to opportunistically activate at runtime if `#include <sys/syscall.h>` could be resolved from Kernel 3.15 or newer.

* I have actually seen this fail in production, once. We lost data.

**Comment by joshudson** [comment]
Can I get someone to check this? I thought I got all the build issues cleaned up but there's one broken build because the build infrastructure faulted.

**Comment by joshudson** [comment]
@jkotas : Somebody needs to look at the failing test cases on the PR. The test cases are actually testing for the behavior being broken on Mac OSX.

**Comment by joshudson** [comment]
Due to changes in the way github works, the branch containing the fix has changed from github.com/joshudson/runtime/master to github.com/joshudson/runtime/master to github.com/joshudson/runtime/dmove
